File: Pythoncode/139.单词拆分.py
class Solution:
    def wordBreak(self, s: str, wordDict) -> bool:
        wordDictSet = set()
        for word in wordDict:
            wordDictSet.add(word)
        dp = {}
        return self.checkBreakable(s, wordDictSet, dp)
    def checkBreakable(self, s, wordDictSet, dp):
        if s in dp:
            return dp[s]
        if len(s) == 0:
            return True
        for i in range(len(s)):
            if self.checkBreakable(s[:i], wordDictSet, dp) and s[i:] in wordDictSet:
                dp[s] = True
                return dp[s]
        dp[s] = False
        return dp[s]

File: Cppcode/139.单词拆分.cpp
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet;
        for (auto s: wordDict) wordSet.insert(s);
        vector<bool> dp(s.size(), false);
        for (auto j = 0; j < dp.size(); j++)
        {
            for (auto i = -1; i < j; i++)
            {
                if (i >= 0 && !dp[i]) continue;
                if (wordSet.count(s.substr(i+1, j-i)) != 0)
                {
                    dp[j] = true;
                    break;
                }
            }
        }
        return dp.back();
    }
};